sklearn/ensemble/_hist_gradient_boosting/splitting.pyx

```diff
@@ -638,18 +727,43 @@ cdef inline Y_DTYPE_C _split_gain(
     """
     cdef:
         Y_DTYPE_C gain
-    gain = negative_loss(sum_gradient_left, sum_hessian_left,
-                         l2_regularization)
-    gain += negative_loss(sum_gradient_right, sum_hessian_right,
-                          l2_regularization)
-    gain -= negative_loss_current_node
+        Y_DTYPE_C value_left
+        Y_DTYPE_C value_right
+
+    # Compute values of potential left and right children
+    value_left = compute_node_value(sum_gradient_left, sum_hessian_left,
+                               lower_bound, upper_bound,
+                               l2_regularization)
+    value_right = compute_node_value(sum_gradient_right, sum_hessian_right,
+                                lower_bound, upper_bound,
+                                l2_regularization)
+
+    if ((monotonic_cst == MonotonicConstraint.POS and value_left > value_right) or
+            (monotonic_cst == MonotonicConstraint.NEG and value_left < value_right)):
+        # don't consider this split since it does not respect the monotonic
+        # constraints. Note that these comparisons need to be done on values
+        # that are already bounded.
+        return -1
+
+    gain = loss_current_node # without bounds
+    gain -= _loss_from_value(value_left, sum_gradient_left)  # with bounds
+    gain -= _loss_from_value(value_right, sum_gradient_right)  # with bounds
+    # Note that the losses for both children are computed with bounded values,
+    # while the loss of the current node isn't. It's OK since all the gain
+    # comparisons will be made with the same loss_current_node anyway.
```

Comment:
Isn't that a problem when using the `min_gain_to_split` hyper-parameter? Does lightgbm has the same problem?

Reply:
I'm reasonably confident that lightgbm does it similarly: see https://github.com/microsoft/LightGBM/blob/master/src/treelearner/feature_histogram.hpp#L160, where the gain of the current node to be split will not take constraints into account (follow `BeforeNumercal` and then `GetLeafGain`). But I could be missing something.

You're right that it doesn't work well with`min_gain_to_split`. I guess the correct way would be to pass the current node's value into `find_node_split`.

I'll try tomorrow and will report back!